adwords_client/client.py

```diff
@@ -276,24 +275,107 @@ def _batch_operations(self, file_name):
+                        if regular_operations:
+                            results.append(service.custom_mutate(previous_client_id, regular_operations))
+                        label_operations = []
+                        regular_operations = []
+                        adw_ops = []
+                adw_ops.append(adwords_operation)
+                previous_client_id = client_id
+                previous_service_name = service_name
+
+        service = self.service(previous_service_name)
+        label_operations = [adw_op for adw_op in adw_ops if 'labelId' in adw_op['operand']]
+        regular_operations = [adw_op for adw_op in adw_ops if 'labelId' not in adw_op['operand']]
+        if label_operations:
+            results.append(service.custom_mutate_labels(previous_client_id, label_operations))
+        if regular_operations:
+            results.append(service.custom_mutate(previous_client_id, regular_operations))
+
+        self._operations_buffer = None
+        return results
 
     # TODO: this method should instantiate a new class (maybe SyncOperation) and transform the internal functions
     # into instance methods. Also, separate the treatment for each "object_type" into a new method as well.
-    def execute_operations(self, operations_folder):
+    def execute_operations(self, operations_folder=None, sync=False):
         """
         Possible columns in the table:
 
         :param table_name:
         :param batchlog_table:
         :return:
         """
-        logger.info('Running %s...', inspect.stack()[0][3])
-        _, files = self.storage.listdir(operations_folder)
-        files = [[path.join(operations_folder, file)] for file in files]
-        self._client = None
-        self._operations_buffer = None
-        self.map_function(self._batch_operations, files)
+        if sync:
+            return self._sync_operations()
+        else:
+            if not operations_folder:
+                raise ValueError('Async operations must have an operation folder defined.')
+            logger.info('Running %s...', inspect.stack()[0][3])
+            _, files = self.storage.listdir(operations_folder)
+            files = [[path.join(operations_folder, file)] for file in files]
+            self._client = None
+            self._operations_buffer = None
+            self.map_function(self._batch_operations, files)
 
     def get_accounts(self, client_id=None):
-        mcs = ManagedCustomerService(self.client)
-        return {k[1]: v[1] for k, v in mcs.get_customers(client_id)}
+        mcs = self.service('ManagedCustomerService')
+        return {account['name']: account for account in mcs.get_customers(client_id)}
+
+    def get_entities(self, get_internal_operation):
+        service_name = self._get_service_from_object_type(get_internal_operation)
+        service = self.service(service_name)
+        results = service.custom_get(get_internal_operation)
```

Comment:
Right now the class that is transforming the `internal_operation` to `adwords_operations` is the `adwords_api.common.Selector`. This could be improved by taking this logic and transferring it to the `adowrds_api.internal_api.builder` module. This way, we could use the builder to create the final operation and only then send it to adwords, using the same logic as for batch operations. This need a bit of thinking, but would make the code much simpler. As it is, this `custom_get` is "translating" and "running" the operations at the same time thanks to my lack of vision at the time I was writing.